_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+
+For the radiator I chose to make Clojure Lambda which polls endpoints and generates plain old html response with automatic refresh. Because AWS Lambda supports Java is Clojure therefore supported. For generating necessary Java classes I used [Lambada](https://github.com/uswitch/lambada) library. The radiator should be run also locally so I wrote also a simple web server for local usage.
+
+### Generating HTML
+
+Generating HTML is easy with Clojure using [Hiccup](https://github.com/weavejester/hiccup) library. Hiccup syntax is also very readable way to make HTML. The main function of the radiator is quite simple.
+
+```clojure
+(defn generate
+  []
+  (hiccup/html
+    [:html
+     [:head
+      [:meta {:charset    "UTF-8"
+              :http-equiv "refresh" :content "10"}]
+      [:meta {:http-equiv "Cache-Control" :content= "no-store"}]
+      [:title config/title]
+      styles]
+     [:body
+      page-header
+      (content)]]))
+```
+One limitation of the serverless approach is that all data for rendering the page should be returned in one response or be linked to outside resources. Lambda is not a web server so there is no way to serve static resources like CSS files. One could of course use S3 or another approach for static files.
+
+### Mix some styles and images
+
+Instead of linking to CSS files I chose to embed necessary CSS in HTML response. For this I used [Garden](https://github.com/noprompt/garden) library which makes CSS generation quite similar like using Hiccup.
+
+The next code snippet shows part of the CSS for the radiator.
+
+```clojure
+(def styles
+  [:style
+    (css [:body {:margin  "0px"
+                 :padding "0px"}])
+    (css [:h1 {:font-family    "proxima-nova,sans-serif"
+               :text-transform "uppercase"
+               :margin         "0px"
+               :font-size      "18px"}])])
+```
+
+The Radiator needs of course some fancy images to visually show when something has gone wrong - alert icons etc. Because HTML images are simply links one can link images from anywhere in Internet. GitHub version of the radiator have preconfigured links to Tango Desktop icons which work quite nicely. The images can be changed via configuration.
+
+### Configuring endpoints
+
+The configuration of the endpoints is done with a simple data structure.
+
+```clojure
+(def projects
+  [{:name             "example account"
+    :aws              {:uri     "https://uri-to.com/api-gateway"
+                       :api-key "secret-api-key"}
+    }])
+```
+
+### Supporting different CI providers
+
+The Radiator should of course support also another CI providers like [GitLab](https://about.gitlab.com/). To make this easier I made a generic data structure for pipelines, alarms and metrics. The radiator should therefore first fetch data from endpoints, after that transform them to generic format which is then rendered to Radiator view.
```

Comment:
"should of course also support"

"transform them into *a* generic format which is then rendered into a Radiator view"